### PR TITLE
Min max on iterables

### PR DESCRIPTION
**Pull request by rjenc29**
Initial commit for CI.  Addresses #3776

### CODE REVIEW

**Comment by rjenc29** [comment]
Well, that was somewhat less successful than I'd hoped.

I thought I'd tweak the `overload` implementation a little to support iterable types, but this breaks some of the parallel implementations.

Perhaps tweaking the `overload` implementation isn't such a good idea.  Let me know what you think if you get a mo.

**Comment by stuartarchibald** [comment]
Thanks for starting work on this fix. The code is largely failing because of a mismatch introduced between the typing and implementation signatures. Typing is declared here: https://github.com/numba/numba/blob/72241de838ca55b97b525dee1a2a83c8bd8e3e42/numba/targets/builtins.py#L477
one implementation has a matching signature:
https://github.com/numba/numba/blob/72241de838ca55b97b525dee1a2a83c8bd8e3e42/numba/targets/builtins.py#L466
the other does not:
https://github.com/numba/numba/blob/72241de838ca55b97b525dee1a2a83c8bd8e3e42/numba/targets/builtins.py#L458

This section of the build log reports the issue, https://travis-ci.org/numba/numba/jobs/501035162#L3792-L3822, feedback on whether this is sufficiently useful is welcome (it's relatively new as a feature).

**Comment by rjenc29** [comment]
My apologies, it's obvious now you point it out.

It's a very useful feature and I am not sure how the logs could be any more useful / direct in this case - I guess as the test failure came from an unexpected place, I must have assumed something more sinister at work.  Please could we just chalk this up to my daftness (there's a pattern developing here...).

Anyway. the last commit is in the right ballpark - I need to sort out the tests and tidy up a bit and then I will un-WIP and invite for a review.  Should be in the next few days.

Cheers

File: numba/tests/test_builtins.py [hunk 1]
```diff
@@ -982,6 +982,60 @@ def test_pow_usecase(self):
...
+            yield lambda: op([np.nan, -np.inf, np.inf, np.nan])
+            yield lambda: op([(1,), (1,), (1,)])
+
+        for fn in sample_functions(op=min):
+            self._check_min_max(fn)
+
+        for fn in sample_functions(op=max):
+            self._check_min_max(fn)
+
+    def test_min_max_supplemental(self):
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py]: Is this covered already by https://github.com/numba/numba/blob/1bfb050f4d69806e7c3b86881f51f21cfbf5ead3/numba/tests/test_parfors.py#L1032-L1084 ?

**Comment by stuartarchibald** [comment]
No problem, it's a new thing! It seems like this PR is along the right lines, thanks for continuing to work on it.

File: numba/tests/test_builtins.py (hunk 1 above)
Comment (rjenc29) [reply numba/tests/test_builtins.py]: It is - I put this test in whilst I was debugging and working through the code.  I intend to take it back out before un-WIP-ing...

Comment (stuartarchibald) [reply numba/tests/test_builtins.py]: No worries, was just checking to save de-duplicating later :)

**Comment by rjenc29** [comment]
If you get a mo, could you let me know how you feel about this incarnation, TA!

File: numba/targets/builtins.py [hunk 2]
```diff
@@ -446,22 +446,48 @@ def impl_index_value(context, builder, sig, args):
...
-        def min_impl(indval1, indval2):
+        def impl(indval1, indval2=None):
             if indval1.value > indval2.value:
                 return indval2
             return indval1
-        return min_impl
+        return impl
+
+    if indval2 is None:
+        if isinstance(indval1, types.IterableType):
```
Comment (stuartarchibald) [inline numba/targets/builtins.py]: Wondering if it's possible to pull these out of this scope which has a non-intuitive signature? Would defining another overload like:
```python
@overload(min)
def iterable_min(iterable):
    if isinstance(iterable, types.IterableType):
        def impl(iterable):
            <details>
```
work? Could also close over the `<`, `>` operator to save duplication too?

File: numba/tests/test_builtins.py [hunk 3]
```diff
@@ -982,6 +982,37 @@ def test_pow_usecase(self):
...
+
+        def sample_functions(op):
+            yield lambda: op(range(10))
+            yield lambda: op(range(4, 12))
+            yield lambda: op(range(-4, -15, -1))
+            yield lambda: op([6.6, 5.5, 7.7])
+            yield lambda: op([(3, 4), (1, 2)])
+            yield lambda: op(frange(1.1, 3.3, 0.1))
+            yield lambda: op([np.nan, -np.inf, np.inf, np.nan])
+            yield lambda: op([(1,), (1,), (1,)])
```
Comment (stuartarchibald) [inline numba/tests/test_builtins.py]: Perhaps vary the tuple value?

File: numba/targets/builtins.py (hunk 2 above)
Comment (rjenc29) [reply numba/targets/builtins.py]: Done.

File: numba/tests/test_builtins.py (hunk 3 above)
Comment (rjenc29) [reply numba/tests/test_builtins.py]: Done.

**Comment by stuartarchibald** [comment]
Thanks for the fixes, this seems good. Whilst the impl does not close over the comparison operation, the llvm inliner should manage to inline the declared function with no trouble.

**Pull request merged**